modules/core/src/main/scala/sangria/execution/ValueCoercionHelper.scala

```diff
@@ -580,107 +580,126 @@ class ValueCoercionHelper[Ctx](
     nodeLocation.toList ++ firstValue.toList
   }
 
-  def isValidValue[In](tpe: InputType[_], input: Option[In])(implicit
-      um: InputUnmarshaller[In]): Vector[Violation] = (tpe, input) match {
-    case (OptionInputType(ofType), Some(value)) if um.isDefined(value) =>
-      isValidValue(ofType, Some(value))
-    case (OptionInputType(_), _) => Vector.empty
-    case (_, None) => Vector(NotNullValueIsNullViolation(sourceMapper, Nil))
-
-    case (ListInputType(ofType), Some(values)) if um.isListNode(values) =>
-      um.getListValue(values)
-        .toVector
-        .flatMap(v =>
-          isValidValue(
-            ofType,
-            v match {
-              case opt: Option[In @unchecked] => opt
-              case other => Option(other)
-            }).map(ListValueViolation(0, _, sourceMapper, Nil)))
-
-    case (ListInputType(ofType), Some(value)) if um.isDefined(value) =>
-      isValidValue(
-        ofType,
-        value match {
-          case opt: Option[In @unchecked] => opt
-          case other => Option(other)
-        }).map(ListValueViolation(0, _, sourceMapper, Nil))
-
-    case (objTpe: InputObjectType[_], Some(valueMap)) if um.isMapNode(valueMap) =>
-      val unknownFields = um.getMapKeys(valueMap).toVector.collect {
-        case f if !objTpe.fieldsByName.contains(f) =>
-          UnknownInputObjectFieldViolation(
-            SchemaRenderer.renderTypeName(objTpe, true),
-            f,
-            sourceMapper,
-            Nil)
-      }
+  def isValidValue[In](inputType: InputType[_], input: Option[In], errorsLimit: Option[Int])(
+      implicit um: InputUnmarshaller[In]): Vector[Violation] = {
+
+    // keeping track of the number of errors
+    var errors = 0
+    def addViolation(violation: Violation): Vector[Violation] = {
+      errors += 1
+      Vector(violation)
+    }
 
-      val fieldViolations =
-        objTpe.fields.toVector.flatMap(f =>
-          isValidValue(f.fieldType, um.getMapValue(valueMap, f.name))
-            .map(MapValueViolation(f.name, _, sourceMapper, Nil)))
+    def isValidValueRec(tpe: InputType[_], in: Option[In])(implicit
```

Comment:
I created a separate recursion function (basically the old method migrated here for the most part) to be called internally so it has easy access to mutable variable